docs/reference/hacl/hash/index.md

```diff
@@ -1,14 +1,27 @@
 # Hash
 
-Hash algorithms in HACL Packages can be used through two APIs: a (simple) one-shot API, and a (more flexible) streaming API.
-
-If you only handle complete messages, i.e., messages that always contain all bytes you want to hash, you can use the one-shot API to easily calculate the digest of that message.
-When you are receiving a message chunk-by-chunk, it may be more efficient to use the streaming API.
-The streaming API has roughly three phases: init, update, and finish.
-You create a context element, call update as often as you need, and then finalize (and cleanup) to obtain a digest.
+Hash algorithms in HACL Packages can be used through two APIs: a (simple) **One-shot API**, and a (more flexible) **Streaming API**.
 
 Note: HACL Packages supports MD5 but does not document it here because it should not be used.
 
+**One-shot API**
+
+If you want to hash a complete messages, i.e., can provide a slice of memory that already contains all the bytes you want to hash, you can use the one-shot API to easily calculate the digest of that message.
+
+**Streaming API**
+
+If you need to provide a message chunk-by-chunk, e.g., because you read the message from network or similar, it could be more appropriate to use the streaming API.
+
+The streaming API has roughly three phases: `init`, `update`, and `finish`.
+Typically, you create a state element by using `init`, call `update` as often as needed, and then call `finish` to obtain the final digest.
+
+**Streaming API (with intermediate digests)**
+
+It is also possible to obtain all intermediate digests by calling `finish` more than once.
+You can call `update("Hello, ")`, and `finish` to obtain the hash of `"Hello, "`.
+Then you can call `update("World!")`, and `finish` *again* to obtain the hash of `"Hello, World!"`.
+There is no need to reinitialize the hash state to obtain both digests.
```

Comment:
Being so explicit about here it should probably be mentioned that `free` needs to be called separately after calling `finish`.